python/qemu/ot/verilator/executer.py

```diff
@@ -0,0 +1,654 @@
+                                                   for t in self._threads)))
+                        break
+                    thread = self._threads.pop(0)
+                    thread.join(timeout=0.2)
+                    if thread.is_alive():
+                        self._threads.append(thread)
+                # retrieve the remaining log messages
+                while log_q:
+                    err, qline = log_q.popleft()
+                    level = logging.ERROR if err else logging.INFO
+                    qline = self._parse_verilator_output(qline, err)
+                    if qline:
+                        self._vlog.log(level, qline)
+                for msg, logger in zip(proc.communicate(timeout=0.1),
+                                       (self._vlog.info, self._vlog.error)):
+                    # should have been captured by the logger threads
+                    for line in msg.splitlines():
+                        line = line.rstrip()
+                        if line:
+                            logger(line)
+                if log_q:
+                    # should never happen
+                    self._vlog.error('Lost traces')
+                self._save_exec_log()
+            tmp_profile = joinpath(workdir, 'profile.vlt')
+            if isfile(tmp_profile) and profile_file:
+                self._log.info('Saving profile file as %s', profile_file)
+                copyfile(tmp_profile, profile_file)
+        return abs(ret or 0)
+
+    def _vtor_logger(self, stream: TextIOWrapper, queue: deque, err: bool) \
+            -> None:
+        # worker thread, blocking on VM stdout/stderr
+        if not stream:
+            self._log.error('No Verilator std%s stream',
+                            'err' if err else 'out')
+            return
+        while self._resume:
+            try:
+                line = stream.readline().rstrip()
+            # pylint: disable=broad-except
+            except Exception as exc:
+                # do not raise an error if stream disappears
+                self._log.error('%s', exc)
+                if self._debug:
+                    print(format_exc(chain=False), file=stderr)
+                break
+            if line:
+                queue.append((err, line))
+            else:
+                sleep(0.005)
+        self._log.debug('End of std%s logger thread', 'err' if err else 'out')
+
+    def _load_devices(self) -> None:
+        args = [self._verilator, '--meminit=list']
+        out = ''
+        with Popen(args, bufsize=1, cwd=self._fm.tmp_dir, encoding='utf-8',
+                   errors='ignore', text=True, stdout=PIPE) as proc:
+            proc.wait(self.START_TIMEOUT)
+            out = proc.stdout.read()
+        mems = re.findall(r"'(\w+)'.*\[(0x[0-9a-f]+),\s*(0x[0-9a-f]+)\]", out)
+        self._devices.update({m[0]: (s:=int(m[1], 16), (int(m[2], 16) - s + 1))
+                             for m in mems})
```

Comment:
Nit: it's up to you, but I think this could do with being made more readable, as it currently requires you to understand the format of values being parsed out from the meminit list.
```suggestion
        for (name, start, end) in mems:
            start = int(start, 16)
            size = int(end, 16) - start + 1
            self._devices[name] = (start, size)
```

Reply:
Ok, I'll also added named groups and a namedtuple